Add Protobuf generation to the codegen utility

JoelSpeed opened the pull request:
This PR adds protobuf as a generator to the codegen generator utility.
The upstream generator unfortunately encodes a lot of logic within the cmd so the `Run` which we are using uses `log.Fatal` a lot when there are errors to return. For now, rather than duplicating the huge run fuction, I think we can work around that.

I've set it up so that the protobuf generator is opt-in (which is different to our other generators) and that it will error if the dependent binaries are missing, but you can make that just a warning with `PROTO_OPTIONAL=1` as you could before, and it will print this in the error messages.

One thing that's new here, is the ability to disable a specific version of the API from generation, this is needed because the `image` group has proto generation on some versions, but not others. We may want to include this in other generators later but for now, it seems only to be needed on this generator.

JoelSpeed committed:
Bootstrap initial go-to-protobuf generator in codegen

JoelSpeed committed:
Add gogoproto to vendor

JoelSpeed committed:
Implement protobuf generator within codegen

JoelSpeed committed:
Add verification of generated protobuf files

JoelSpeed committed:
Enable codegen protobuf on legacy APIs

JoelSpeed committed:
Add the ability to skip protobuf generation for certain versions

JoelSpeed committed:
Update verify protobuf scripts

elmiko reviewed (commented):
code generally looks good to me, although i admit i'm not an expert on the go/protobuf tooling. i like all the comments, thanks for adding those.

i tried testing this by running `make update-codegen` from the root of a clean project, but i get this error

```
topological order github.com/openshift/api/apps/v1
2023/01/13 14:37:23 /home/mike/dev/openshift-api/github.com/openshift/api/apps/v1/generated.proto: Input is shadowed in the --proto_path by "github.com/openshift/api/apps/v1/generated.proto".  Either use the latter file as your input or reorder the --proto_path so that the former file's location comes first.
2023/01/13 14:37:23 protoc -I . -I /home/mike/dev/openshift-api -I /home/mike/dev/openshift-api/vendor --gogo_out=/home/mike/dev/openshift-api /home/mike/dev/openshift-api/github.com/openshift/api/apps/v1/generated.proto
2023/01/13 14:37:23 Unable to generate protoc on github.com.openshift.api.apps.v1: exit status 1
```

did i do something wrong?

JoelSpeed commented:
Interesting error! I don't think you did something wrong, but, are you using GOPATH? I believe that the generator doesn't currently work if you aren't in the standard GOPATH structure

JoelSpeed commented:
> /home/mike/dev/openshift-api/github.com/openshift/api/apps/v1

This isn't the right GOPATH structure expected by the tooling, you need `$GOPATH/src/github/openshift/api` as the path for the API checkout, can you try moving it here?

elmiko commented:
i can try moving the openshift/api directory into my gopath, but i usually work on stuff outside the gopath. fwiw, my go path is like `$HOME/go`.

elmiko commented:
running from within the gopath did work

JoelSpeed commented:
> running from within the gopath did work

Unfortunately, I know the protobuf generators aren't currently working outside of GOPATH, so I guess I could add a check in that warns about this 🤔  If you have it set up as expected, great, if you don't it causes errors as you've seen. Some of our other generators have a fix in for this but I'm waiting on upstream approval to fix the protobuf generator

elmiko commented:
if the check isn't too annoying to write i'd say sure, otherwise i think it would be fine to just print a warning message when that command starts. then at least i know what i did wrong.

danwinship commented:
/uncc

JoelSpeed commented:
/lifecycle frozen

JoelSpeed commented:
/remove-lifecycle stale

JoelSpeed commented:
@elmiko I've rebased this now and added two further commits, one updates the dependencies for the tools module, the second adds `TrimPathPrefix` as a config for the `go-to-protobuf` generator. Just tested, and it now works with `GOPATH=` at the beginning, so I reckon we don't need those warnings anymore

elmiko commented:
i tried this out on my local machine, outside of the GOPATH, by running `make update-protobuf` but i get an error, did i mess something up?

```
$ make update-protobuf
hack/update-protobuf.sh
make[1]: Entering directory '/home/mike/dev/openshift-api/tools'
Tools vendor version mismatch, removing old utils
Writing tools vendor version e633c236d76c2053897f2bc7a16abc47986cd2f0
go build -mod=vendor -o /home/mike/dev/openshift-api/tools/_output/bin/linux/amd64/protoc-gen-gogo ./vendor/k8s.io/code-generator/cmd/go-to-protobuf/protoc-gen-gogo
make[1]: Leaving directory '/home/mike/dev/openshift-api/tools'
make[1]: Entering directory '/home/mike/dev/openshift-api/tools'
Building codegen version 4894dd2bb191564ce2c1cb52d3132f4ae8a7365b
make[1]: Leaving directory '/home/mike/dev/openshift-api/tools'
I0811 11:26:24.894736  154102 root.go:80] Running generators for 
I0811 11:26:24.894781  154102 root.go:80] Running generators for apiserver.openshift.io
I0811 11:26:24.894787  154102 root.go:80] Running generators for apps.openshift.io
I0811 11:26:24.894843  154102 generator.go:113] Generating protobuf functions for for apps.openshift.io/v1
2023/08/11 11:26:26 protoc -I . -I /home/mike/dev/openshift-api -I /home/mike/dev/openshift-api/vendor --gogo_out=/home/mike/dev/openshift-api /home/mike/dev/openshift-api/github.com/openshift/api/apps/v1/generated.proto
2023/08/11 11:26:26 /home/mike/dev/openshift-api/github.com/openshift/api/apps/v1/generated.proto: Input is shadowed in the --proto_path by "github.com/openshift/api/apps/v1/generated.proto".  Either use the latter file as your input or reorder the --proto_path so that the former file's location comes first.

2023/08/11 11:26:26 Unable to generate protoc on github.com.openshift.api.apps.v1: exit status 1
make: *** [Makefile:91: update-protobuf] Error 1
```

it also appears to drop a `github.com` directory in my top-level repository directory.

JoelSpeed commented:
/test all

elmiko commented:
i saw you made an update here and gave it another run locally, different error this time:

```
$ make update-protobuf
hack/update-protobuf.sh
make[1]: Entering directory '/home/mike/dev/openshift-api/tools'
make[1]: Leaving directory '/home/mike/dev/openshift-api/tools'
make[1]: Entering directory '/home/mike/dev/openshift-api/tools'
make[1]: Leaving directory '/home/mike/dev/openshift-api/tools'
I0814 15:22:31.681216  636451 root.go:80] Running generators for 
I0814 15:22:31.681246  636451 root.go:80] Running generators for apiserver.openshift.io
I0814 15:22:31.681249  636451 root.go:80] Running generators for apps.openshift.io
I0814 15:22:31.681291  636451 generator.go:113] Generating protobuf functions for for apps.openshift.io/v1
2023/08/14 15:22:33 protoc -I . -I /home/mike/dev/openshift-api -I /home/mike/dev/openshift-api/vendor --gogo_out=/home/mike/dev/openshift-api apps/v1/generated.proto
2023/08/14 15:22:33 google/protobuf/descriptor.proto: File not found.
github.com/gogo/protobuf/gogoproto/gogo.proto:32:1: Import "google/protobuf/descriptor.proto" was not found or had errors.
github.com/gogo/protobuf/gogoproto/gogo.proto:38:8: "google.protobuf.EnumOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.EnumOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.EnumOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.EnumOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.EnumOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto:46:8: "google.protobuf.EnumValueOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto:50:8: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FileOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto:92:8: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.MessageOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto:129:8: "google.protobuf.FieldOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FieldOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FieldOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FieldOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FieldOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FieldOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FieldOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FieldOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FieldOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FieldOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FieldOptions" is not defined.
github.com/gogo/protobuf/gogoproto/gogo.proto: "google.protobuf.FieldOptions" is not defined.
apps/v1/generated.proto:8:1: Import "github.com/gogo/protobuf/gogoproto/gogo.proto" was not found or had errors.

2023/08/14 15:22:33 Unable to generate protoc on github.com.openshift.api.apps.v1: exit status 1
make: *** [Makefile:91: update-protobuf] Error 1
```

JoelSpeed commented:
/remove-lifecycle rotten

JoelSpeed commented on tools/codegen/pkg/protobuf/protobuf.go:
```diff
@@ -0,0 +1,229 @@
+
+	// Check the pre-requisite binaries exist.
+	if err := checkBinaries(); err != nil {
+		if os.Getenv("PROTO_OPTIONAL") != "" {
+			klog.Warningf("Skipping protobuf generation: %v", err)
+			return nil
+		}
```
Discussed sync, need to account for GOPATH and, if not in GOPATH, also skip. Should move proto optional to skip everything even if you have the binaries

JoelSpeed commented:
/remove-lifecycle stale

JoelSpeed committed:
Move pre-flight checks earlier to start of genGroup

everettraven commented on tools/codegen/pkg/generation/types.go:
```diff
@@ -23,6 +23,10 @@ type Config struct {
 	// When omitted, the default configuration will be used.
 	OpenAPI *OpenAPIConfig `json:"openapi,omitempty"`
 
+	// Prootbuf represents the configuration of the go-to-protobuf generator.
```
```suggestion
	// Protobuf represents the configuration of the go-to-protobuf generator.
```

everettraven reviewed (commented):
Overall this LGTM.

Some of the "Major" comments from coderabbit seem like they would be reasonable to address.

JoelSpeed commented on tools/codegen/cmd/go_to_protobuf.go:
```diff
@@ -0,0 +1,46 @@
+func init() {
+	rootCmd.AddCommand(goToProtobufCmd)
+
+	rootCmd.PersistentFlags().StringVar(&goToProtobufHeaderFilePath, "go-to-protobuf:header-file-path", "", "Path to file containing boilerplate header text. The string YEAR will be replaced with the current 4-digit year. When omitted, no header is added to the generated files.")
```
This is a feature of the upstream generator, which will handle this for us, we do not need to substitute the year ourselves

JoelSpeed commented:
@everettraven All issues coderabbit reported should now be resolved

everettraven reviewed (approved):
/lgtm

JoelSpeed commented:
/verified by @JoelSpeed
